File: euler66.pl
#!/usr/bin/perl -w
use strict;
use bigint;

my ($ans, $result)=(0, 0);
my ($n, $a, $t, $b, $flag);
my ($kkk, $kk, $k, $hhh, $hh, $h);

foreach (1..1000) {
    $n = int(sqrt($_));
    next if ($n**2==$_);
    ($a, $t, $b, $flag) = ($n, $n, 1, 1);
    ($kkk, $kk, $hhh, $hh) = (1, 0, 0, 1);
    while ($flag) {
        $k = $kkk+$t*$kk;
        $h = $hhh+$t*$hh;
        if ($h**2-$_*$k**2==1) {
            $flag = 0;
            #print $h**2 ."-". $_ ."*". $k**2 ."=1\n";
        }
        ($kkk, $kk, $hhh, $hh) = ($kk, $k, $hh, $h);
        $t = int($b*($n+$a)/($_-$a**2));
        $b = ($_-$a*$a)/$b;
        $a = int(($n+$a)/$b)*$b-$a;
    }
    if ($result<$h) {
        $result = $h;
        $ans = $_;
    }
}

print "$ans $result\n";

File: euler66.py
#!/usr/bin/python

ans = 0
result = 0

for i in range(1, 1001):
    n = int(i**0.5)
    if (n**2!=i):
        a = n
        t = n
        b = 1
        flag = 1
        kkk = 1
        kk = 0
        hhh = 0
        hh = 1
        while (flag):
            k = kkk+t*kk
            h = hhh+t*hh
            if (h**2-i*k**2==1):
                flag = 0
            kkk = kk
            kk = k
            hhh = hh
            hh = h
            t = int(b*(n+a)/(i-a**2))
            b = (i-a**2)/b
            a = int((n+a)/b)*b-a
        if (result < h):
            result = h
            ans = i

print ans, result
